src/Aspire.Hosting/ApplicationModel/ExpressionResolver.cs

```diff
@@ -0,0 +1,127 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Globalization;
+
+namespace Aspire.Hosting.ApplicationModel;
+
+internal class ExpressionResolver(string containerHostName, CancellationToken cancellationToken)
+{
+    class HostAndPortPresence
+    {
+        public bool HasHost { get; set; }
+        public bool HasPort { get; set; }
+    }
+
+    // For each endpoint, keep track of whether host and port are in use
+    // The key is the unique name of the endpoint, which is the resource name and endpoint name
+    readonly Dictionary<string, HostAndPortPresence> _endpointUsage = [];
+    static string EndpointUniqueName(EndpointReference endpointReference) => $"{endpointReference.Resource.Name}/{endpointReference.EndpointName}";
+
+    // This marks whether we are in the preprocess phase or not
+    // Not thread-safe, but we doesn't matter, since this class is never used concurrently
+    bool Preprocess { get; set; }
+
+    async Task<string?> EvalEndpointAsync(EndpointReference endpointReference, EndpointProperty property)
+    {
+        // In the preprocess phase, our only goal is to determine if the host and port properties are both used
+        // for each endpoint.
+        if (Preprocess)
+        {
+            if (!_endpointUsage.TryGetValue(EndpointUniqueName(endpointReference), out var hostAndPortPresence))
+            {
+                hostAndPortPresence = new HostAndPortPresence();
+                _endpointUsage[EndpointUniqueName(endpointReference)] = hostAndPortPresence;
+            }
+
+            if (property is EndpointProperty.Host or EndpointProperty.IPV4Host)
+            {
+                hostAndPortPresence.HasHost = true;
+            }
+            else if (property == EndpointProperty.Port)
+            {
+                hostAndPortPresence.HasPort = true;
+            }
+            else if (property == EndpointProperty.Url)
+            {
+                hostAndPortPresence.HasHost = hostAndPortPresence.HasPort = true;
+            }
+
+            return string.Empty;
+        }
+
+        // We need to use the root resource, e.g. AzureStorageResource instead of AzureBlobResource
+        // Otherwise, we get the wrong values for IsContainer and Name
+        var target = endpointReference.Resource.GetRootResource();
+
+        bool HasBothHostAndPort() =>
+            _endpointUsage[EndpointUniqueName(endpointReference)].HasHost &&
+            _endpointUsage[EndpointUniqueName(endpointReference)].HasPort;
+
+        return (property, target.IsContainer()) switch
```

Comment:
This could be part of the pattern match:

```suggestion
        bool HasBothHostAndPort() =>
            _endpointUsage[EndpointUniqueName(endpointReference)].HasHost &&
            _endpointUsage[EndpointUniqueName(endpointReference)].HasPort;

        return (property, target.IsContainer(), HasBothHostAndPort()) switch
```